src/pressure/VCycleMG.cpp

```diff
@@ -753,154 +612,244 @@ void VCycleMG::Solve(Field &out, Field &tmp, Field const &b, size_t level, bool
         // TODO Error handling
     }
 
-    const real dx = domain->get_dx(out.get_level());
-    const real dy = domain->get_dy(out.get_level());
-    const real dz = domain->get_dz(out.get_level());
+// Diffusion step
+    (this->*m_solve_diffusion_function)(out, tmp, b, level, sync);
+    if (sync) {
+#pragma acc wait
+    }
+}
 
-    auto params = Parameters::getInstance();
+void VCycleMG::call_smooth_colored_gauss_seidel(Field &out, Field &tmp, Field const &b, const size_t level, bool sync) {
+    auto domain = Domain::getInstance();
+
+    // local variables and parameters for GPU
+    const real dx = domain->get_dx(level);
+    const real dy = domain->get_dy(level);
+    const real dz = domain->get_dz(level);
 
     FieldType type = out.get_type();
+    real *data_out = out.data;
 
-    auto boundary = BoundaryController::getInstance();
+    BoundaryController *boundary = BoundaryController::getInstance();
+    boundary->apply_boundary(data_out, level, type, sync);
 
-    size_t *d_iList = boundary->get_inner_list_level_joined();
-    size_t *d_bList = boundary->get_boundary_list_level_joined();
+    const real rdx2 = 1. / (dx * dx);
+    const real rdy2 = 1. / (dy * dy);
+    const real rdz2 = 1. / (dz * dz);
 
-    auto bsize_i = boundary->get_size_inner_list_level_joined();
-    auto bsize_b = boundary->get_size_boundary_list_level_joined();
+    // preparation for diffusion step (alpha, beta)
+    const real alphaX = rdx2;
+    const real alphaY = rdy2;
+    const real alphaZ = rdz2;
 
-    // start/ end
+    const real rbeta = 2. * (alphaX + alphaY + alphaZ);
+    const real beta = 1. / rbeta;
+
+    tmp.copy_data(out);
+#pragma acc data present(out, tmp, b)
+    {
+        for (int i = 0; i < m_n_relax; i++) {
+            ColoredGaussSeidelDiffuse::colored_gauss_seidel_step(out, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
+            boundary->apply_boundary(data_out, level, type, sync); // for res/err only Dirichlet BC
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247